src/include/storage/metric/thread_level_stats_collector.h

```diff
@@ -0,0 +1,372 @@
+                          catalog::namespace_oid_t namespace_oid, catalog::table_oid_t table_oid) {
+    for (auto &metric : metric_dispatch_[StatsEventType::TUPLE_INSERT])
+      metric->OnTupleInsert(txn, database_oid, namespace_oid, table_oid);
+  }
+
+  /**
+   * Collector action on tuple delete
+   * @param txn Context of the transaction performing delete
+   * @param database_oid OID of the database that the tuple delete happens
+   * @param namespace_oid OID of the namespace that the tuple delete happens
+   * @param table_oid OID of the table that the tuple delete happens
+   */
+  void CollectTupleDelete(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid,
+                          catalog::namespace_oid_t namespace_oid, catalog::table_oid_t table_oid) {
+    for (auto &metric : metric_dispatch_[StatsEventType::TUPLE_DELETE])
+      metric->OnTupleDelete(txn, database_oid, namespace_oid, table_oid);
+  }
+
+  /**
+   * Collector action on table memory allocation
+   * @param database_oid OID of the database that the table memory allocation happens
+   * @param namespace_oid OID of the namespace that the table memory allocation happens
+   * @param table_oid OID of the table that the table memory allocation happens
+   * @param size number of bytes being allocated
+   */
+  void CollectTableMemoryAlloc(catalog::db_oid_t database_oid, catalog::namespace_oid_t namespace_oid,
+                               catalog::table_oid_t table_oid, size_t size) {
+    for (auto &metric : metric_dispatch_[StatsEventType::TABLE_MEMORY_ALLOC])
+      metric->OnTableMemoryAlloc(database_oid, namespace_oid, table_oid, size);
+  }
+
+  /**
+   * Collector action on table memory free
+   * @param database_oid OID of the database that the table memory free happens
+   * @param namespace_oid OID of the namespace that the table memory free happens
+   * @param table_oid OID of the table that the table memory free happens
+   * @param size number of bytes being freed
+   */
+  void CollectTableMemoryFree(catalog::db_oid_t database_oid, catalog::namespace_oid_t namespace_oid,
+                              catalog::table_oid_t table_oid, size_t size) {
+    for (auto &metric : metric_dispatch_[StatsEventType::TABLE_MEMORY_FREE])
+      metric->OnTableMemoryFree(database_oid, namespace_oid, table_oid, size);
+  }
+
+  /**
+   * Collector action on table memory usage
+   * @param database_oid OID of the database that the table memory usage happens
+   * @param namespace_oid OID of the namespace that the table memory usage happens
+   * @param table_oid OID of the table that the table memory usage happens
+   * @param size number of bytes being used
+   */
+  void CollectTableMemoryUsage(catalog::db_oid_t database_oid, catalog::namespace_oid_t namespace_oid,
+                               catalog::table_oid_t table_oid, size_t size) {
+    for (auto &metric : metric_dispatch_[StatsEventType::TABLE_MEMORY_ALLOC])
```

Comment:
Is `TABLE_MEMORY_ALLOC` used for a reason here? In the `StatsEventType` enum there's an `INDEX_MEMORY_USAGE` but no `TABLE_MEMORY_USAGE`

Reply:
Corrected